src/scan/postgres_scan.cpp

```diff
@@ -78,99 +68,161 @@ PostgresScanGlobalState::InitGlobalState(duckdb::TableFunctionInitInput &input)
+	named_parameters["cardinality"] = duckdb::LogicalType::UBIGINT;
+	named_parameters["relid"] = duckdb::LogicalType::UINTEGER;
+	named_parameters["snapshot"] = duckdb::LogicalType::POINTER;
+	projection_pushdown = true;
+	filter_pushdown = true;
+	filter_prune = true;
+	cardinality = PostgresScanCardinality;
+}
 
-	auto relid = PostgresFunctionGuard(FindMatchingRelation, schema_name, table_name);
-	if (relid == InvalidOid) {
-		return nullptr;
+duckdb::unique_ptr<duckdb::GlobalTableFunctionState>
+PostgresScanTableFunction::PostgresScanInitGlobal(duckdb::ClientContext &, duckdb::TableFunctionInitInput &input) {
+	auto &bind_data = input.bind_data->CastNoConst<PostgresScanFunctionData>();
+	auto global_state = duckdb::make_uniq<PostgresScanGlobalState>(bind_data.snapshot, bind_data.rel, input);
+	return global_state;
+}
+
+duckdb::unique_ptr<duckdb::LocalTableFunctionState>
+PostgresScanTableFunction::PostgresScanInitLocal(duckdb::ExecutionContext &, duckdb::TableFunctionInitInput &,
+                                                 duckdb::GlobalTableFunctionState *gstate) {
+	auto global_state = reinterpret_cast<PostgresScanGlobalState *>(gstate);
+	return duckdb::make_uniq<PostgresScanLocalState>(global_state);
+}
+void
+SetOutputCardinality(duckdb::DataChunk &output, PostgresScanLocalState &local_state) {
+	idx_t output_cardinality =
+	    local_state.output_vector_size <= STANDARD_VECTOR_SIZE ? local_state.output_vector_size : STANDARD_VECTOR_SIZE;
+	output.SetCardinality(output_cardinality);
+	local_state.output_vector_size -= output_cardinality;
+}
+
+void
+PostgresScanTableFunction::PostgresScanFunction(duckdb::ClientContext &, duckdb::TableFunctionInput &data,
+                                                duckdb::DataChunk &output) {
+	auto &local_state = data.local_state->Cast<PostgresScanLocalState>();
+
+	/* We have exhausted table scan */
+	if (local_state.exhausted_scan) {
+		SetOutputCardinality(output, local_state);
+		return;
 	}
 
-	auto tuple = PostgresFunctionGuard(SearchSysCache1, RELOID, ObjectIdGetDatum(relid));
-	if (!HeapTupleIsValid(tuple)) {
-		elog(WARNING, "(PGDuckDB/PostgresReplacementScan) Cache lookup failed for relation %u", relid);
-		return nullptr;
+	local_state.output_vector_size = 0;
+
+	int i = 0;
+	for (; i < STANDARD_VECTOR_SIZE; i++) {
+		TupleTableSlot *slot = local_state.global_state->table_reader_global_state->GetNextTuple();
+		if (TupIsNull(slot)) {
+			local_state.exhausted_scan = true;
+			break;
+		}
+		slot_getallattrs(slot);
+		InsertTupleIntoChunk(output, local_state, slot);
 	}
 
-	auto relForm = (Form_pg_class)GETSTRUCT(tuple);
-	if (relForm->relkind != RELKIND_VIEW) {
-		PostgresFunctionGuard(ReleaseSysCache, tuple);
-		return nullptr;
+	/* If we finish before reading complete vector means that scan was exhausted. */
+	if (i != STANDARD_VECTOR_SIZE) {
+		local_state.exhausted_scan = true;
```

Comment:
Isn't it already done L207? And since the only other way to exit the loop above is to have `i==STANDARD_VECTOR_SIZE` this branch is redundant no?